File: src/sim/solar.ts
import type { LatLon } from '../types/geo'
import type { Site } from '../types/site'
import type { SolarPositionSample, SolarPositionSeries } from '../types/weather'
import type { Celsius, Degrees, EpochMillis, Meters, Millibars, Radians } from '../types/units'
import { physicsCore, requirePhysicsCore } from './core'
import { clamp, cosDeg, normaliseDegrees, RAD_TO_DEG, sinDeg, tanDeg } from './math'
import { FIELDS_PER_SAMPLE as RUST_FIELDS_PER_SAMPLE } from './rust-core'
import { SOLAR_CONSTANT_W_M2 } from './units'

export interface SpaObserver {
  readonly location: LatLon
  readonly elevationM: Meters
  readonly pressureMb: Millibars
  readonly temperatureC: Celsius
}

export type SolarPositionAlgorithm = 'nrel-spa' | 'psa' | 'grena3' | 'michalsky'

const SUN_DISC_HALF_WIDTH_DEG = 0.26667
const ATMOSPHERIC_REFRACTION_DEG = 0.5667
export const refractionCorrectionDeg = (
  geometricElevationDeg: Degrees,
  pressureMb: Millibars,
  temperatureC: Celsius,
): Degrees => {
  if (geometricElevationDeg < -(SUN_DISC_HALF_WIDTH_DEG + ATMOSPHERIC_REFRACTION_DEG)) {
    return 0 as Degrees
  }
  // SPA / Bennett 1982
  return (((pressureMb / 1010) * (283 / (273 + temperatureC)) * 1.02) /
    (60 * tanDeg(geometricElevationDeg + 10.3 / (geometricElevationDeg + 5.11)))) as Degrees
}

// Kasten & Young 1989
export const kastenYoungAirMass = (
  zenithDeg: Degrees,
  pressureMb: Millibars,
): { readonly relative: number; readonly absolute: number } => {
  if (zenithDeg >= 90) return { relative: 0, absolute: 0 }
  const relative = 1 / (cosDeg(zenithDeg) + 0.50572 * (96.07995 - zenithDeg) ** -1.6364)
  return { relative, absolute: relative * (pressureMb / 1013.25) }
}

export const pressureFromElevation = (elevationM: Meters): Millibars =>
  (1013.25 * Math.exp(-elevationM / 8434.5)) as Millibars

export const SEA_LEVEL_M = 0 as Meters

/**
 * The single place a site's elevation enters the solar chain. Where the lookup returned
 * nothing the observer sits at the sea-level reference and the site readout says the
 * elevation is unknown, rather than reporting an assumed 0 m as a measurement
 */
export const observerFor = (site: Site): SpaObserver => {
  const elevationM = site.elevationM ?? SEA_LEVEL_M
  return {
    location: site.location,
    elevationM,
    pressureMb: pressureFromElevation(elevationM),
    temperatureC: (site.normals.monthlyMeanTempC[6] ?? 12) as Celsius,
  }
}

export const extraterrestrialNormal = (earthRadiusVectorAu: number): number =>
  earthRadiusVectorAu > 0 ? SOLAR_CONSTANT_W_M2 / earthRadiusVectorAu ** 2 : 0

export const angleOfIncidenceCos = (
  zenithRad: Radians,
  solarAzimuthRad: Radians,
  surfaceTiltRad: Radians,
  surfaceAzimuthRad: Radians,
): number =>
  Math.cos(zenithRad) * Math.cos(surfaceTiltRad) +
  Math.sin(zenithRad) * Math.sin(surfaceTiltRad) * Math.cos(solarAzimuthRad - surfaceAzimuthRad)

// tan psi = tan(elevation) / |cos(gamma_s - gamma_c)|; sigma records which side of the row the sun is on
export const profileAngle = (
  solarElevationRad: Radians,
  solarAzimuthRad: Radians,
  rowAzimuthRad: Radians,
): { readonly psiRad: Radians; readonly side: -1 | 0 | 1 } => {
  const cosDelta = Math.cos(solarAzimuthRad - rowAzimuthRad)
  if (Math.abs(cosDelta) < 1e-12) return { psiRad: (Math.PI / 2) as Radians, side: 0 }
  const side: -1 | 1 = cosDelta > 0 ? 1 : -1
  return {
    psiRad: Math.atan2(Math.tan(solarElevationRad), Math.abs(cosDelta)) as Radians,
    side,
  }
}

// ENU frame: +x East, +y North, +z Up (doc 03 section 0)
export const sunUnitVector = (
  elevationDeg: Degrees,
  azimuthDeg: Degrees,
): { readonly x: number; readonly y: number; readonly z: number } => {
  const cosEl = cosDeg(elevationDeg)
  return { x: cosEl * sinDeg(azimuthDeg), y: cosEl * cosDeg(azimuthDeg), z: sinDeg(elevationDeg) }
}

/**
 * NREL's Solar Position Algorithm, computed by `crates/agv-sim`.
 *
 * The 160 lines of periodic-term summation that used to be here, and the 321 rows of coefficients
 * in `spa-tables.ts` beside them, are gone: `crates/agv-sim/src/solar.rs` is the implementation
 * now and `crates/agv-sim/tests/nrel_spa.rs` holds it to NREL/TP-560-34302 Appendix A.5. This is
 * a single-sample call over the batched ABI, which is the shape `state/sun.ts` wants for one
 * instant and the reason the batched entry point is also exported.
 *
 * It refuses rather than falling back. There is no second implementation to fall back TO, and one
 * that produced plausible numbers from a different algorithm would be worse than none.
 */
export const spaPosition = (utcMillis: EpochMillis, observer: SpaObserver): SolarPositionSample => {
  const flat = requirePhysicsCore().spaSeriesFlat([utcMillis], {
    latitudeDeg: observer.location.latitudeDeg,
    longitudeDeg: observer.location.longitudeDeg,
    elevationM: observer.elevationM,
    pressureMb: observer.pressureMb,
    temperatureC: observer.temperatureC,
  })
  const read = (index: number): number => flat[index] ?? Number.NaN
  return {
    geometricElevationDeg: read(0) as Degrees,
    apparentElevationDeg: read(1) as Degrees,
    zenithDeg: read(2) as Degrees,
    azimuthDeg: read(3) as Degrees,
    declinationDeg: read(4) as Degrees,
    hourAngleDeg: read(5) as Degrees,
    earthRadiusVectorAu: read(6),
    relativeAirMass: read(7),
    absoluteAirMass: read(8),
  }
}

const michalskyPosition = (utcMillis: EpochMillis, observer: SpaObserver): SolarPositionSample => {
  const jd = utcMillis / 86_400_000 + 2_440_587.5
  const n = jd - 2451545
  const meanLongitude = normaliseDegrees(280.46 + 0.9856474 * n)
  const meanAnomaly = normaliseDegrees(357.528 + 0.9856003 * n)
  const lambda = meanLongitude + 1.915 * sinDeg(meanAnomaly) + 0.02 * sinDeg(2 * meanAnomaly)
  const epsilon = 23.439 - 0.0000004 * n
  const alpha = normaliseDegrees(
    RAD_TO_DEG * Math.atan2(cosDeg(epsilon) * sinDeg(lambda), cosDeg(lambda)),
  )
  const declination = RAD_TO_DEG * Math.asin(clamp(sinDeg(epsilon) * sinDeg(lambda), -1, 1))
  const gmstHours = (18.697374558 + 24.06570982441908 * n) % 24
  const hourAngleDeg = normaliseDegrees(gmstHours * 15 + observer.location.longitudeDeg - alpha)
  const latitude = observer.location.latitudeDeg
  const geometricElevationDeg =
    RAD_TO_DEG *
    Math.asin(
      clamp(
        sinDeg(latitude) * sinDeg(declination) +
          cosDeg(latitude) * cosDeg(declination) * cosDeg(hourAngleDeg),
        -1,
        1,
      ),
    )
  const apparentElevationDeg =
    geometricElevationDeg +
    refractionCorrectionDeg(
      geometricElevationDeg as Degrees,
      observer.pressureMb,
      observer.temperatureC,
    )
  const azimuthDeg = normaliseDegrees(
    180 +
      RAD_TO_DEG *
        Math.atan2(
          sinDeg(hourAngleDeg),
          cosDeg(hourAngleDeg) * sinDeg(latitude) - tanDeg(declination) * cosDeg(latitude),
        ),
  )
  const zenithDeg = 90 - apparentElevationDeg
  const radiusAu = 1.00014 - 0.01671 * cosDeg(meanAnomaly) - 0.00014 * cosDeg(2 * meanAnomaly)
  const airMass = kastenYoungAirMass(zenithDeg as Degrees, observer.pressureMb)
  return {
    geometricElevationDeg: geometricElevationDeg as Degrees,
    apparentElevationDeg: apparentElevationDeg as Degrees,
    zenithDeg: zenithDeg as Degrees,
    azimuthDeg: azimuthDeg as Degrees,
    declinationDeg: declination as Degrees,
    hourAngleDeg: hourAngleDeg as Degrees,
    earthRadiusVectorAu: radiusAu,
    relativeAirMass: airMass.relative,
    absoluteAirMass: airMass.absolute,
  }
}

export const solarPositionSeries = (
  utcMillis: Float64Array,
  observer: SpaObserver,
  algorithm: SolarPositionAlgorithm,
): SolarPositionSeries => {
  if (algorithm === 'psa' || algorithm === 'grena3') {
    throw new Error(`solar position algorithm '${algorithm}' is not implemented: use 'nrel-spa'`)
  }
  const count = utcMillis.length
  const geometricElevationDeg = new Float32Array(count)
  const apparentElevationDeg = new Float32Array(count)
  const azimuthDeg = new Float32Array(count)
  const absoluteAirMass = new Float32Array(count)
  const extraterrestrialNormalWM2 = new Float32Array(count)

  /*
    The Rust core answers for the SPA only. Michalsky is a different algorithm, deliberately kept
    as a cheap cross-check on the SPA rather than ported beside it, so asking for it here has to
    keep running the TypeScript whatever is installed. Getting that wrong would mean the check and
    the thing it checks were the same code
  */
  const core = algorithm === 'michalsky' ? null : physicsCore()
  if (core !== null) {
    const flat = core.spaSeriesFlat(utcMillis, {
      latitudeDeg: observer.location.latitudeDeg,
      longitudeDeg: observer.location.longitudeDeg,
      elevationM: observer.elevationM,
      pressureMb: observer.pressureMb,
      temperatureC: observer.temperatureC,
    })
    for (let i = 0; i < count; i += 1) {
      const at = i * RUST_FIELDS_PER_SAMPLE
      geometricElevationDeg[i] = flat[at] ?? Number.NaN
      apparentElevationDeg[i] = flat[at + 1] ?? Number.NaN
      azimuthDeg[i] = flat[at + 3] ?? Number.NaN
      absoluteAirMass[i] = flat[at + 8] ?? Number.NaN
      extraterrestrialNormalWM2[i] = extraterrestrialNormal(flat[at + 6] ?? 0)
    }
  } else {
    const evaluate = algorithm === 'michalsky' ? michalskyPosition : spaPosition
    for (let i = 0; i < count; i += 1) {
      const sample = evaluate((utcMillis[i] ?? 0) as EpochMillis, observer)
      geometricElevationDeg[i] = sample.geometricElevationDeg
      apparentElevationDeg[i] = sample.apparentElevationDeg
      azimuthDeg[i] = sample.azimuthDeg
      absoluteAirMass[i] = sample.absoluteAirMass
      extraterrestrialNormalWM2[i] = extraterrestrialNormal(sample.earthRadiusVectorAu)
    }
  }
  return {
    count,
    geometricElevationDeg,
    apparentElevationDeg,
    azimuthDeg,
    absoluteAirMass,
    extraterrestrialNormalWM2,
  }
}

File: crates/agv-sim/src/solar.rs
//! NREL's Solar Position Algorithm, ported from `src/sim/solar.ts`.
//!
//! Reda & Andreas 2008, NREL/TP-560-34302. The periodic terms in `spa_tables.rs` are generated
//! from the TypeScript by `scripts/generate-rust-tables.mjs` rather than retyped, so there is one
//! copy of 364 lines of coefficients and a drift check on it.
//!
//! The port is deliberately literal. Where the TypeScript reads awkwardly the Rust reads
//! awkwardly in the same place, because the thing being protected is agreement with an
//! implementation that has been checked against the published worked example, not elegance.

use crate::math::{clamp, cos_deg, normalise_degrees, sin_deg, tan_deg, DEG_TO_RAD, RAD_TO_DEG};
use crate::spa_tables::{
    B0, B1, L0, L1, L2, L3, L4, L5, NUTATION_ABCD_ARRAY, NUTATION_YTERM_ARRAY, R0, R1, R2, R3, R4,
};
use crate::time::{delta_t_seconds, julian_time};

const SUN_DISC_HALF_WIDTH_DEG: f64 = 0.26667;
const ATMOSPHERIC_REFRACTION_DEG: f64 = 0.5667;
const EARTH_RADIUS_M: f64 = 6_378_140.0;

/// Where the observer is and what the air above them is doing.
#[derive(Debug, Clone, Copy)]
pub struct Observer {
    pub latitude_deg: f64,
    pub longitude_deg: f64,
    pub elevation_m: f64,
    pub pressure_mb: f64,
    pub temperature_c: f64,
}

/// Every figure `spaPosition` returns, in the same order and under the same names.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SolarPosition {
    pub geometric_elevation_deg: f64,
    pub apparent_elevation_deg: f64,
    pub zenith_deg: f64,
    pub azimuth_deg: f64,
    pub declination_deg: f64,
    pub hour_angle_deg: f64,
    pub earth_radius_vector_au: f64,
    pub relative_air_mass: f64,
    pub absolute_air_mass: f64,
}

fn sum_mult_cos_add_mult(terms: &[[f64; 3]], x: f64) -> f64 {
    terms.iter().map(|t| t[0] * (t[1] + t[2] * x).cos()).sum()
}

/// Horner over the periodic-term series, outermost first, exactly as the TypeScript does it.
fn polynomial(series: &[&[[f64; 3]]], jme: f64) -> f64 {
    let mut total = 0.0;
    for terms in series.iter().rev() {
        total = total * jme + sum_mult_cos_add_mult(terms, jme);
    }
    total
}

fn heliocentric_longitude_deg(jme: f64) -> f64 {
    normalise_degrees(polynomial(&[L0, L1, L2, L3, L4, L5], jme) / 1e8 * RAD_TO_DEG)
}

fn heliocentric_latitude_deg(jme: f64) -> f64 {
    polynomial(&[B0, B1], jme) / 1e8 * RAD_TO_DEG
}

fn heliocentric_radius_au(jme: f64) -> f64 {
    polynomial(&[R0, R1, R2, R3, R4], jme) / 1e8
}

fn nutation(jce: f64) -> (f64, f64) {
    let x = [
        297.85036 + 445267.11148 * jce - 0.0019142 * jce.powi(2) + jce.powi(3) / 189_474.0,
        357.52772 + 35999.05034 * jce - 0.0001603 * jce.powi(2) - jce.powi(3) / 300_000.0,
        134.96298 + 477198.867398 * jce + 0.0086972 * jce.powi(2) + jce.powi(3) / 56_250.0,
        93.27191 + 483202.017538 * jce - 0.0036825 * jce.powi(2) + jce.powi(3) / 327_270.0,
        125.04452 - 1934.136261 * jce + 0.0020708 * jce.powi(2) + jce.powi(3) / 450_000.0,
    ];
    let mut psi = 0.0;
    let mut epsilon = 0.0;
    for (y, abcd) in NUTATION_YTERM_ARRAY.iter().zip(NUTATION_ABCD_ARRAY.iter()) {
        let arg = (0..5).map(|i| y[i] * x[i]).sum::<f64>() * DEG_TO_RAD;
        psi += (abcd[0] + abcd[1] * jce) * arg.sin();
        epsilon += (abcd[2] + abcd[3] * jce) * arg.cos();
    }
    (psi / 36_000_000.0, epsilon / 36_000_000.0)
}

fn mean_ecliptic_obliquity_arcsec(jme: f64) -> f64 {
    let u = jme / 10.0;
    84381.448 - 4680.93 * u - 1.55 * u.powi(2) + 1999.25 * u.powi(3)
        - 51.38 * u.powi(4)
        - 249.67 * u.powi(5)
        - 39.05 * u.powi(6)
        + 7.12 * u.powi(7)
        + 27.87 * u.powi(8)
        + 5.79 * u.powi(9)
        + 2.45 * u.powi(10)
}

/// SPA / Bennett 1982.
pub fn refraction_correction_deg(
    geometric_elevation_deg: f64,
    pressure_mb: f64,
    temperature_c: f64,
) -> f64 {
    if geometric_elevation_deg < -(SUN_DISC_HALF_WIDTH_DEG + ATMOSPHERIC_REFRACTION_DEG) {
        return 0.0;
    }
    (pressure_mb / 1010.0) * (283.0 / (273.0 + temperature_c)) * 1.02
        / (60.0 * tan_deg(geometric_elevation_deg + 10.3 / (geometric_elevation_deg + 5.11)))
}

/// Kasten & Young 1989, the path length at sea-level pressure.
///
/// This is the argument the Perez transposition wants. Pressure-correcting it is what DISC and
/// DIRINT want and the two are not interchangeable: see `TranspositionInput::relative_air_mass`.
pub fn relative_air_mass(zenith_deg: f64) -> f64 {
    if zenith_deg >= 90.0 {
        return 0.0;
    }
    1.0 / (cos_deg(zenith_deg) + 0.50572 * (96.07995 - zenith_deg).powf(-1.6364))
}

/// Kasten & Young 1989, relative and pressure-corrected.
pub fn kasten_young_air_mass(zenith_deg: f64, pressure_mb: f64) -> (f64, f64) {
    let relative = relative_air_mass(zenith_deg);
    (relative, relative * (pressure_mb / 1013.25))
}

pub fn pressure_from_elevation(elevation_m: f64) -> f64 {
    1013.25 * (-elevation_m / 8434.5).exp()
}

pub fn spa_position(utc_millis: f64, observer: &Observer) -> SolarPosition {
    let jt = julian_time(utc_millis, delta_t_seconds(utc_millis));
    let jme = jt.julian_ephemeris_millennium;
    let jce = jt.julian_ephemeris_century;

    let radius_au = heliocentric_radius_au(jme);
    let theta = normalise_degrees(heliocentric_longitude_deg(jme) + 180.0);
    let beta_geo = -heliocentric_latitude_deg(jme);

    let (delta_psi, delta_epsilon) = nutation(jce);
    let epsilon = mean_ecliptic_obliquity_arcsec(jme) / 3600.0 + delta_epsilon;
    let lambda = theta + delta_psi - 20.4898 / (3600.0 * radius_au);

    let nu0 = normalise_degrees(
        280.46061837
            + 360.98564736629 * (jt.julian_day - 2_451_545.0)
            + 0.000387933 * jt.julian_century.powi(2)
            - jt.julian_century.powi(3) / 38_710_000.0,
    );
    let nu = nu0 + delta_psi * cos_deg(epsilon);

    let alpha = normalise_degrees(
        RAD_TO_DEG
            * (sin_deg(lambda) * cos_deg(epsilon) - tan_deg(beta_geo) * sin_deg(epsilon))
                .atan2(cos_deg(lambda)),
    );
    let declination = RAD_TO_DEG
        * clamp(
            sin_deg(beta_geo) * cos_deg(epsilon)
                + cos_deg(beta_geo) * sin_deg(epsilon) * sin_deg(lambda),
            -1.0,
            1.0,
        )
        .asin();

    let latitude = observer.latitude_deg;
    let longitude = observer.longitude_deg;
    let hour_angle_deg = normalise_degrees(nu + longitude - alpha);

    let xi = 8.794 / (3600.0 * radius_au);
    let u = (0.99664719 * tan_deg(latitude)).atan();
    let x_term = u.cos() + (observer.elevation_m / EARTH_RADIUS_M) * cos_deg(latitude);
    let y_term = 0.99664719 * u.sin() + (observer.elevation_m / EARTH_RADIUS_M) * sin_deg(latitude);

    let delta_alpha = RAD_TO_DEG
        * (-x_term * sin_deg(xi) * sin_deg(hour_angle_deg))
            .atan2(cos_deg(declination) - x_term * sin_deg(xi) * cos_deg(hour_angle_deg));
    let declination_prime = RAD_TO_DEG
        * ((sin_deg(declination) - y_term * sin_deg(xi)) * cos_deg(delta_alpha))
            .atan2(cos_deg(declination) - x_term * sin_deg(xi) * cos_deg(hour_angle_deg));
    let hour_angle_prime = hour_angle_deg - delta_alpha;

    let geometric_elevation_deg = RAD_TO_DEG
        * clamp(
            sin_deg(latitude) * sin_deg(declination_prime)
                + cos_deg(latitude) * cos_deg(declination_prime) * cos_deg(hour_angle_prime),
            -1.0,
            1.0,
        )
        .asin();
    let apparent_elevation_deg = geometric_elevation_deg
        + refraction_correction_deg(
            geometric_elevation_deg,
            observer.pressure_mb,
            observer.temperature_c,
        );

    // atan2 form is mandatory: the single-argument form mirrors afternoon shadows (doc 03 s1.5)
    let azimuth_deg = normalise_degrees(
        180.0
            + RAD_TO_DEG
                * sin_deg(hour_angle_prime).atan2(
                    cos_deg(hour_angle_prime) * sin_deg(latitude)
                        - tan_deg(declination_prime) * cos_deg(latitude),
                ),
    );

    let zenith_deg = 90.0 - apparent_elevation_deg;
    let (relative, absolute) = kasten_young_air_mass(zenith_deg, observer.pressure_mb);

    SolarPosition {
        geometric_elevation_deg,
        apparent_elevation_deg,
        zenith_deg,
        azimuth_deg,
        declination_deg: declination_prime,
        hour_angle_deg: hour_angle_prime,
        earth_radius_vector_au: radius_au,
        relative_air_mass: relative,
        absolute_air_mass: absolute,
    }
}

/// The direction of the sun as a unit vector in the site's local frame.
///
/// X east, Y north, Z up, matching `geom.rs`. Returned as a tuple rather than a `UnitVec3` so
/// this module keeps no dependency on the geometry types: solar position is upstream of geometry
/// and stays that way.
pub fn sun_unit_vector(elevation_deg: f64, azimuth_deg: f64) -> (f64, f64, f64) {
    let cos_el = cos_deg(elevation_deg);
    (
        cos_el * sin_deg(azimuth_deg),
        cos_el * cos_deg(azimuth_deg),
        sin_deg(elevation_deg),
    )
}

/// The sun's elevation measured in the plane perpendicular to a row, and which side it is on.
///
/// This is the angle row-to-row shading is decided by: a sun high in the sky but far along the row
/// axis casts a shadow that misses the next row entirely, and the profile angle is what expresses
/// that. `side` is `Math.sign`'s three-valued answer and not a boolean, because zero means the sun
/// is exactly along the row and neither side is shaded.
pub fn profile_angle(
    solar_elevation_rad: f64,
    solar_azimuth_rad: f64,
    row_azimuth_rad: f64,
) -> (f64, f64) {
    let cos_delta = (solar_azimuth_rad - row_azimuth_rad).cos();
    if cos_delta.abs() < 1e-12 {
        return (std::f64::consts::FRAC_PI_2, 0.0);
    }
    let side = if cos_delta > 0.0 { 1.0 } else { -1.0 };
    (solar_elevation_rad.tan().atan2(cos_delta.abs()), side)
}
